Fix sanitizer issues

Hello newcomer

Since 702cd0cfd025c313e5e165af5394a83fcf01959a one can easily enable [undefinied behavior sanitizer](http://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html).

Simply use clang as toolchain and use debug/asan (e.g. by adding `-DENABLE_DEBUG=ON -DENABLE_ASAN=ON -DCMAKE_C_COMPILER=clang-3.8 -DCMAKE_CXX_COMPILER=clang++-3.8` to your cmake invocation). It might work for gcc, too.

Then try to run all tests (see tree.libelektra.org/doc/TESTING.md) and fix some issues that are reported. Do not try to fix all at once, but create a PR as early as possible.

See also:
https://build.libelektra.org/jenkins/job/elektra-clang-asan/lastFailedBuild/console
https://build.libelektra.org/jenkins/job/elektra-gcc-asan/lastFailedBuild/console

The error messages would be interesting. If lua/java cause the problems we might have to exclude the tests (at least for the version/OS where it does not work).

I get now a segmentation fault I had not earlier (only with ASAN):
```
[ 13%] Generating exported_symbols.h
Segmentation fault
src/libs/elektra/CMakeFiles/elektra-core.dir/build.make:53: recipe for target 'src/libs/elektra/exported_symbols.h' failed
```

Unfortunately, no core dump gets generated, even though they are activated. Have to investigate it later.

--- a/doc/COMPILE.md
+++ b/doc/COMPILE.md
@@ -11,7 +11,7 @@ Or on RPM based systems (CentOS):
 
 	sudo yum install -y cmake3 gcc-c++
 
-Or on macOS Sierra, most of the build tools can be obtained by installing XCode (from the App Store). Other required tools may be installed using [brew](http://brew.sh/). First install brew as described on their website. Then issue the following command to get cmake to complete the basic requirements:
+Or on macOS Sierra, most of the build tools can be obtained by installing Xcode (from the App Store). Other required tools may be installed using [brew](http://brew.sh/). First install brew as described on their website. Then issue the following command to get cmake to complete the basic requirements:
 
 	brew install cmake
 

--- a/scripts/sed
+++ b/scripts/sed
@@ -19,6 +19,7 @@ s/\<[Ll]ibreoffice/LibreOffice/g
 s/\<mark[Dd]own\>/Markdown/g
 
 s/\<unix|UNIX/Unix/g
+s/\<XCode\>/Xcode/g
 
 # ===================
 # = Technical Terms =

--- a/src/libs/loader/dl.c
+++ b/src/libs/loader/dl.c
@@ -53,9 +53,9 @@ int elektraModulesInit (KeySet * modules, Key * error ELEKTRA_UNUSED)
 elektraPluginFactory elektraModulesLoad (KeySet * modules, const char * name, Key * errorKey)
 {
 #ifdef _WIN32
-	const char elektraPluginPostfix[] = ".dll";
+	static const char elektraPluginPostfix[] = ".dll";
 #else
-	const char elektraPluginPostfix[] = ".so";
+	static const char elektraPluginPostfix[] = ".so";
 #endif
 
 	Key * moduleKey = keyNew ("system/elektra/modules", KEY_END);

--- a/src/plugins/crypto/botan_operations.cpp
+++ b/src/plugins/crypto/botan_operations.cpp
@@ -9,10 +9,12 @@
 
 #include <botan/auto_rng.h>
 #include <botan/filters.h>
+#include <botan/hmac.h>
 #include <botan/init.h>
 #include <botan/lookup.h>
 #include <botan/pbkdf2.h>
 #include <botan/pipe.h>
+#include <botan/sha2_32.h>
 #include <botan/symkey.h>
 #include <kdbplugin.h>
 #include <memory>
@@ -70,8 +72,8 @@ static int getKeyIvForEncryption (KeySet * config, Key * errorKey, Key * masterK
 		const kdb_unsigned_long_t iterations = CRYPTO_PLUGIN_FUNCTION (getIterationCount) (errorKey, config);
 
 		// generate/derive the cryptographic key and the IV
-		unique_ptr<PBKDF> pbkdf = unique_ptr<PBKDF> (get_pbkdf ("PBKDF2(SHA-256)"));
-		OctetString derived = pbkdf->derive_key (
+		PKCS5_PBKDF2 pbkdf (new HMAC (new SHA_256));
+		OctetString derived = pbkdf.derive_key (
 			requiredKeyBytes, std::string (reinterpret_cast<const char *> (keyValue (masterKey)), keyGetValueSize (masterKey)),
 			salt, sizeof (salt), iterations);
 
@@ -121,8 +123,8 @@ static int getKeyIvForDecryption (KeySet * config, Key * errorKey, Key * masterK
 	try
 	{
 		// derive the cryptographic key and the IV
-		unique_ptr<PBKDF> pbkdf = unique_ptr<PBKDF> (get_pbkdf ("PBKDF2(SHA-256)"));
-		OctetString derived = pbkdf->derive_key (
+		PKCS5_PBKDF2 pbkdf (new HMAC (new SHA_256));
+		OctetString derived = pbkdf.derive_key (
 			requiredKeyBytes, std::string (reinterpret_cast<const char *> (keyValue (masterKey)), keyGetValueSize (masterKey)),
 			saltBuffer, saltBufferLen, iterations);
 

--- a/src/plugins/crypto/gcrypt_operations.c
+++ b/src/plugins/crypto/gcrypt_operations.c
@@ -324,7 +324,7 @@ int elektraCryptoGcryEncrypt (elektraCryptoHandle * handle, Key * k, Key * error
 	// encrypt the value using gcrypt's in-place encryption
 	const size_t dataLen =
 		outputLen - ELEKTRA_CRYPTO_GCRY_BLOCKSIZE - sizeof (kdb_unsigned_long_t) - saltLen - ELEKTRA_CRYPTO_MAGIC_NUMBER_LEN;
-	memcpy (current, content, contentLen);
+	if (contentLen) memcpy (current, content, contentLen);
 	gcry_err = gcry_cipher_encrypt (*handle, current, dataLen, NULL, 0);
 	if (gcry_err != 0)
 	{

--- a/tests/sanitizer.blacklist
+++ b/tests/sanitizer.blacklist
@@ -1,1 +1,7 @@
 src:.*boost/smart_ptr/detail/.*
+# memory leaks reported by asan due to xerces interfaces, no issue according to valgrind
+src:*xerces/*.cpp
+src:*xerces/*.hpp
+# memory leaks reported by asan in botan, no issue according to valgrind
+src:*crypto/botan_operations.cpp
+src:*botan/pbkdf.h

--- a/tests/shell/shell_recorder/tutorial_wrapper/CMakeLists.txt
+++ b/tests/shell/shell_recorder/tutorial_wrapper/CMakeLists.txt
@@ -15,12 +15,10 @@ function (add_plugin_shell_test PLUGIN)
 endfunction ()
 
 add_s_test (tutorial_cascading "${CMAKE_SOURCE_DIR}/doc/tutorials/cascading.md")
-add_s_test (tutorial_validation "${CMAKE_SOURCE_DIR}/doc/tutorials/validation.md")
 add_s_test (kdb-complete "${CMAKE_SOURCE_DIR}/doc/help/kdb-complete.md")
 add_s_test (kdb-global-umount "${CMAKE_SOURCE_DIR}/doc/help/kdb-global-umount.md")
 add_s_test (kdb-ls "${CMAKE_SOURCE_DIR}/doc/help/kdb-ls.md")
 
-add_plugin_shell_test (base64)
 add_plugin_shell_test (blockresolver)
 add_plugin_shell_test (cachefilter)
 add_plugin_shell_test (camel)
@@ -31,15 +29,24 @@ add_plugin_shell_test (ini)
 add_plugin_shell_test (ipaddr)
 add_plugin_shell_test (line)
 add_plugin_shell_test (mathcheck)
-add_plugin_shell_test (mini)
 add_plugin_shell_test (mozprefs)
 add_plugin_shell_test (multifile)
 add_plugin_shell_test (range)
-add_plugin_shell_test (tcl)
 add_plugin_shell_test (type)
 add_plugin_shell_test (xerces)
 add_plugin_shell_test (yamlcpp)
 add_plugin_shell_test (yajl)
+
+# The tests below might fail on Linux if ASAN is enabled because of problematic code in `stl_tree.h`
+# The cause of this seems to be a bug in old versions of `libstdc++`:
+# http://lists.llvm.org/pipermail/cfe-dev/2013-August/031194.html
+set (ASAN_LINUX ENABLE_ASAN AND ${CMAKE_SYSTEM_NAME} MATCHES "Linux")
+if (NOT (ASAN_LINUX AND CMAKE_CXX_COMPILER_ID MATCHES "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 3.6))
+	add_s_test (tutorial_validation "${CMAKE_SOURCE_DIR}/doc/tutorials/validation.md")
+	add_plugin_shell_test (base64)
+	add_plugin_shell_test (mini)
+	add_plugin_shell_test (tcl)
+endif (NOT (ASAN_LINUX AND CMAKE_CXX_COMPILER_ID MATCHES "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 3.6))
 
 # Only works with super user privileges, since it writes to `/etc/hosts`:
 # add_s_test (tutorial_mount "${CMAKE_SOURCE_DIR}/doc/tutorials/mount.md")

--- a/tests/shell/shell_recorder/tutorial_wrapper/CMakeLists.txt
+++ b/tests/shell/shell_recorder/tutorial_wrapper/CMakeLists.txt
@@ -21,6 +21,7 @@ add_s_test (kdb-global-umount "${CMAKE_SOURCE_DIR}/doc/help/kdb-global-umount.md
 add_s_test (kdb-ls "${CMAKE_SOURCE_DIR}/doc/help/kdb-ls.md")
 
 add_plugin_shell_test (base64)
+add_plugin_shell_test (blockresolver)
 add_plugin_shell_test (cachefilter)
 add_plugin_shell_test (camel)
 add_plugin_shell_test (conditionals)
@@ -30,20 +31,15 @@ add_plugin_shell_test (ini)
 add_plugin_shell_test (ipaddr)
 add_plugin_shell_test (line)
 add_plugin_shell_test (mathcheck)
+add_plugin_shell_test (mini)
 add_plugin_shell_test (mozprefs)
+add_plugin_shell_test (multifile)
 add_plugin_shell_test (range)
+add_plugin_shell_test (tcl)
 add_plugin_shell_test (type)
 add_plugin_shell_test (xerces)
+add_plugin_shell_test (yamlcpp)
 add_plugin_shell_test (yajl)
-
-# The tests below might fail on Linux if ASAN is enabled because of problematic code in `stl_tree.h`
-if (NOT (${CMAKE_SYSTEM_NAME} MATCHES "Linux" AND ENABLE_ASAN))
-	add_plugin_shell_test (blockresolver)
-	add_plugin_shell_test (mini)
-	add_plugin_shell_test (multifile)
-	add_plugin_shell_test (tcl)
-	add_plugin_shell_test (yamlcpp)
-endif (NOT (${CMAKE_SYSTEM_NAME} MATCHES "Linux" AND ENABLE_ASAN))
 
 # Only works with super user privileges, since it writes to `/etc/hosts`:
 # add_s_test (tutorial_mount "${CMAKE_SOURCE_DIR}/doc/tutorials/mount.md")

